packages/backend-app-api/src/wiring/BackendInitializer.ts

```diff
@@ -20,9 +20,9 @@ import {
   coreServices,
   ServiceRef,
   ServiceFactory,
+  LifecycleService,
+  RootLifecycleService,
 } from '@backstage/backend-plugin-api';
-import { BackendLifecycleImpl } from '../services/implementations/rootLifecycle/rootLifecycleServiceFactory';
```

Comment:
Could you please clarify why we want to inline these types?

Reply:
This is mentioned in the PR body at the top ^.

It wasn't inlined as such. It's just that app-api ideally should have no business knowing about the specific implementation of these interfaces. The old code had an `instanceof` hack in place just to be able to call the implementation methods for startup and shutdown that were NOT part of the public interface - those only exist on the implementation classes. I made that use "duck typing" instead of `instanceof`. The reason for that is: we now have lifecycle implementations potentially in two different places. Only one of those places could ever match the `instanceof`. By changing to duck typing, I made sure that app-api now can accept any class, as long as it (1) fulfills the public API, and (2) has those two methods.

This is slightly related to the [redactions service discussion](https://github.com/backstage/backstage/pull/24730) - these are two "write-side" methods that we in this case do NOT want to expose to the read-side plugins.

Reply:
Thank you for explaining, TIL 💯